src/drive/web/modules/public/PublicToolbar.jsx

```diff
@@ -21,65 +27,140 @@ const DownloadFilesButton = ({ t, onDownload, size = 'normal' }) => (
+      {t('toolbar.menu_download_folder')}
+    </MenuItem>
   </Menu>
 )
 
-const PublicToolbar = (
-  { files, onDownload, breakpoints: { isMobile }, renderInBar = false },
-  { t }
-) => {
-  if (isMobile) {
-    return (
-      <BarRight>
-        <MoreMenu t={t} onDownload={() => onDownload(files)} />
-      </BarRight>
-    )
-  } else if (renderInBar) {
-    return (
-      <BarRight>
-        <div className={toolbarstyles['toolbar-inside-bar']}>
-          <CozyHomeLink from="sharing-drive" />
-          <DownloadFilesButton
-            t={t}
-            onDownload={() => onDownload(files)}
-            size="small"
-          />
-        </div>
-      </BarRight>
-    )
-  } else {
-    return (
-      <div className={toolbarstyles['fil-toolbar-files']} role="toolbar">
+const MobileToolbar = ({ onDownload, discoveryLink, redirectTo }, { t }) => (
+  <BarRight>
+    <MoreMenu
+      t={t}
+      onDownload={onDownload}
+      onOpenInCozy={discoveryLink ? () => redirectTo(discoveryLink) : false}
+      onCreateCozy={
+        discoveryLink
+          ? false
+          : () => redirectTo(getHomeLinkHref('sharing-drive'))
+      }
+    />
+  </BarRight>
+)
+
+const CozybarToolbar = ({ onDownload, discoveryLink }, { t }) => (
+  <BarRight>
+    <div className={toolbarstyles['toolbar-inside-bar']}>
+      {discoveryLink ? (
+        <OpenInCozyButton href={discoveryLink} t={t} size="small" />
+      ) : (
         <CozyHomeLink from="sharing-drive" />
-        <DownloadFilesButton t={t} onDownload={() => onDownload(files)} />
-        <BarRight>
-          <div />
-        </BarRight>
-      </div>
-    )
+      )}
+      <DownloadFilesButton t={t} onDownload={onDownload} size="small" />
+    </div>
+  </BarRight>
+)
+
+const DesktopToolbar = ({ onDownload, discoveryLink }, { t }) => (
+  <div className={toolbarstyles['fil-toolbar-files']} role="toolbar">
+    {discoveryLink ? (
+      <OpenInCozyButton href={discoveryLink} t={t} />
+    ) : (
+      <CozyHomeLink from="sharing-drive" />
+    )}
+    <DownloadFilesButton t={t} onDownload={onDownload} />
+    <BarRight>
+      <div />
+    </BarRight>
+  </div>
+)
+
+class PublicToolbar extends React.Component {
+  state = {
+    discoveryLink: null
+  }
+
+  componentDidMount() {
+    if (window.location.pathname === '/preview') this.loadSharingDiscoveryLink()
+  }
+
+  async loadSharingDiscoveryLink() {
+    try {
+      const { client } = this.context
+      const response = await client
+        .collection('io.cozy.permissions')
+        .getOwnPermissions()
+      const sourceId = response.data.attributes.source_id
+      const sharingId = sourceId.split('/')[1]
+
+      const { sharecode } = getQueryParameter()
+
+      const link = client
+        .collection('io.cozy.sharings')
+        .getDiscoveryLink(sharingId, sharecode)
+      this.setState({ discoveryLink: link })
+    } catch (err) {
+      console.warn('Failed to load sharing discovery link', err)
+    }
+  }
```

Comment:
Not a big fan of having these requests and manipulations inside a component. Is there an easy way to move it outside?

Reply:
As we discussed, one way to do it might be to use redux for this. It's a bit of a pain though, as we'd have to add dedicated actions, reducers and a store section for this view, which isn't even in the main app. Given the extra code bloat and fragmentation that would introduce, I'm honestly not sure it's worth it.

We could also transform the function into a standalone function, it would take `client` as a parameter and return the link / `null`. The call would still be in the component, but at least the code itself would be a simple js function ?